### PR TITLE
Add support for proxqp and osqp solver

### PR DESCRIPTION
**Pull request by fabinsch**
Hello, with this PR I would like to propose a way to integrate **PROXQP** and **OSQP** solver into TSID. I added options to the CMakeLists file to support the solvers, which will first check if the [proxsuite](https://github.com/simple-robotics/proxsuite) and [osqp-eigen](https://github.com/robotology/osqp-eigen) are available (both can be easily installed via conda for example).

The integration is done in c++ and I added the corresponding python bindings. The unittests, both cpp and python, got modified to test also the new solvers.

In addition to the `HQPData` (holding the constraints and cost as TSID objects), this PR adds a struct called QPData which is gathering all necessary eigen matrices to describe a QP problem. Currently, these matrices are members of each solver.

I am looking forward to your feedback.

### CODE REVIEW

**Comment by nim65s** [comment]
This looks very good, thank you.

Moving eiquadprog to HQPData seems to be the right thing to do, but the current version is a breaking change.

Maybe we should replace them by deprecated properties, instead of removing them ?

**Comment by fabinsch** [comment]
Hi @nim65s, thanks for your review. Ok, so you are saying instead of removing the protected members `m_H`, `m_g`, `m_CI` etc to a new struct, I should leave them where they are and use `TSID_DEPRECATED` in their declaration?

I thought that I could move them as they were protected members of each solver and only used inside the solvers. I took care to not break the API for the user by not changing the function signatures. 

But you are right, a problem could be if someone implemented his own solver that is inheriting from `SolverHQuadProg` for example as the members `m_H`, `m_g` etc. disappear through this PR.

**Review by jcarpent — APPROVED** [review]
Thanks a lot @fabinsch for the nice work.
A small mention of the fact that both OSQP and ProxQP do not require any regularization.

**Comment by fabinsch** [comment]
> Thanks a lot @fabinsch for the nice work. A small mention of the fact that both OSQP and ProxQP do not require any regularization.

Thanks for the clarification. I've not seen a hessian regularization before in proxqp or osqp that's why I set the `bool hessianRegularization` to `false` by default for proxqp and osqp. If you want and think its more consistent, I can remove it entirely from both solvers.

**Comment by fabinsch** [comment]
Hello @andreadelprete, I was wondering if you as the main developer of this repo might have some time to review this PR in the near future by any chance ? Thanks a lot

**Comment by andreadelprete** [comment]
> Hello @andreadelprete, I was wondering if you as the main developer of this repo might have some time to review this PR in the near future by any chance ? Thanks a lot

Yes, I was following this development but could not find the time to review it yet, very bad week at home. Hopefully tomorrow I'll do it.

File: src/solvers/solver-HQP-eiquadprog-fast.cpp [hunk 1]
```diff
@@ -86,105 +87,112 @@ namespace tsid
...
+              m_qpData.g.noalias() -= w*constr->matrix().transpose()*constr->vector();
+            }
+            
+            if (hessianRegularization)
+            {
+              double m_hessian_regularization(DEFAULT_HESSIAN_REGULARIZATION);
+              m_qpData.H.diagonal().array() += m_hessian_regularization;
+            }
+          }
+    }
```
Comment (andreadelprete) [inline src/solvers/solver-HQP-eiquadprog-fast.cpp:179]: There seems to be an indentation issue here, which makes the changes of the PR hard to understand

**Review by andreadelprete — COMMENTED** [review]
This is a really nice PR, with lots of improvements here and there. The only thing I am failing to understand is the benefit of introducing the new data structure QPData. Is it motivated by your need for accessing this data from outside the solvers? Apart from this there were a few small issues, for which I've left specific comments.

File: src/solvers/solver-HQP-eiquadprog-fast.cpp (hunk 1 above)
Comment (fabinsch) [reply src/solvers/solver-HQP-eiquadprog-fast.cpp:179]: Oups sorry, I corrected it.

**Comment by fabinsch** [comment]
Hello @andreadelprete and thanks for your review. I integrated your comments into the PR. 

And yes, the QPData structure was meant to gather the complete package of matrices describing a general qp problem, which can then easily be exposed in python, extracted, and used in different contexts outside the solvers. What do you think about it?

**Comment by nim65s** [comment]
for my part, I think this is a nice feature, but also a breaking change. Maybe we could de-couple this new structure from the addition of those solvers ? I'm pretty sure we are going to have troubles when our users will upgrade TSID and their code will stop working.

**Comment by jcarpent** [comment]
I think it is hard to decouple both.
I would say we can increase the major release version.
We have also in mind to support SOCP problems in addition to QPs, meaning that will depict Coulomb friction cone constraints as a second-order cone.

**Comment by fabinsch** [comment]
hey @nim65s, thanks for sharing your opinion. I am wondering what in this PR is for you the breaking change. I did not change the signature of the functions of the eiquadprog(-fast) solvers, the members are still part of the solver (marked with deprecated) but additionally, we have the new QPData struct, and I added two functions. I think it should still work for the users, but maybe I am missing something. Thanks for your help

**Review by andreadelprete — APPROVED** [review]
This seems ok to me now. Let's see what @nim65s says.

**Comment by nim65s** [comment]
OK for me if we bump the major version :)
I'll take the opportunity to setup the code format & linters before that

**Comment by jcarpent** [comment]
> OK for me if we bump the major version :) I'll take the opportunity to setup the code format & linters before that

As mentioned by @fabinsch, it seems there is no break of API in the current PR.
@nim65s Could you confirm?

**Comment by jcarpent** [comment]
We have a break of API indeed, as mentioned by @nim65s in private messages.
So, let's merge this PR and then remove useless and deprecated quantities.

**Pull request merged**